compiler/util/llvmDebug.cpp

```diff
@@ -0,0 +1,743 @@
+#endif
+            );
+        
+          N = this->dibuilder.createPointerType(
+            pteStrDIType,
+            layout->getPointerSizeInBits(ty->getPointerAddressSpace()),
+            0,
+            name);
+        
+          myTypeDescriptors[type] = N;
+          return toDITYPE(N);
+        }
+      }
+      else if(type->astTag == E_AggregateType) {
+        // dealing with classes
+        AggregateType *this_class = (AggregateType *)type;
+        llvm::SmallVector<LLVM_METADATA_OPERAND_TYPE *, 8> EltTys;
+        LLVM_DITYPE derivedFrom = nullptr;
+        if( type->dispatchParents.length() > 0 )
+          derivedFrom = get_type(type->dispatchParents.first());
+
+        // solve the data class: _ddata
+        if(this_class->symbol->hasFlag(FLAG_DATA_CLASS)) {
+          Type* vt = getDataClassType(this_class->symbol)->typeInfo();
+          if(vt) {
+            N = this->dibuilder.createPointerType(
+              get_type(vt),
+              layout->getPointerSizeInBits(ty->getPointerAddressSpace()),
+              0,
+              name);
+            
+            myTypeDescriptors[type] = N;
+            return toDITYPE(N);
+          }
+        } //Not sure whether we should directly return getType(vt)
+        
+        const llvm::StructLayout* slayout = NULL;
+        const char *struct_name = this_class->classStructName(true);
+        llvm::Type* st = getTypeLLVM(struct_name);
+        if(st){
+          llvm::StructType* struct_type = llvm::cast<llvm::StructType>(st);
+          if(!struct_type->isOpaque()){
+            N = this->dibuilder.createForwardDecl(
+              llvm::dwarf::DW_TAG_structure_type, 
+              name,
+              get_module_scope(defModule),
+              get_file(defFile),
+              defLine,
+              0, // RuntimeLang
+              layout->getTypeSizeInBits(ty),
+              8*layout->getABITypeAlignment(ty));
+     
+            //N is added to the map (early) so that element search below can find it,
+            //so as to avoid infinite recursion for structs that contain pointers to
+            //their own type.
+            myTypeDescriptors[type] = N;
+
+            slayout = layout->getStructLayout(struct_type); 
+            for_fields(field, this_class) {
+              // field is a Symbol
+              const char* fieldDefFile = field->defPoint->fname();
+              int fieldDefLine = field->defPoint->linenum();
+              TypeSymbol* fts = field->type->symbol;
+              llvm::Type* fty = fts->llvmType;
+              LLVM_DITYPE mty;
+              LLVM_DITYPE fditype =  get_type(field->type);
+              if(fditype == NULL)
+              // if field->type is an internal type, get_type returns null
+              // which is not a good type for a MemberType). At the moment it
+              // uses a nullptr type as a stub, but we should change it
+                fditype = this->dibuilder.createNullPtrType();
```

Comment:
Currently get_type returns NULL if a type is internal or if, for some reasons, it can’t understand the correct debug type. This presents a side effect in which sometimes it expects a type and the constructor returns NULL (like when building an object), generating bad informations. At the moment this is solved using this type stub, but for provide better debugging informations it’s necessary a more sophisticated method.